src/Abp.RedisCache/Runtime/Caching/Redis/AbpRedisCache.cs

```diff
@@ -97,11 +97,16 @@ public override void Set(KeyValuePair<string, object>[] pairs, TimeSpan? sliding
                                           (GetLocalizedRedisKey(p.Key), Serialize(p.Value, GetSerializableType(p.Value)))
                                          );
 
+            _database.StringSet(redisPairs.ToArray());
+
             if (slidingExpireTime.HasValue || absoluteExpireTime.HasValue)
             {
-                Logger.WarnFormat("{0}/{1} is not supported for Redis bulk insert of key-value pairs", nameof(slidingExpireTime), nameof(absoluteExpireTime));
+                var expireTime = absoluteExpireTime ?? slidingExpireTime ?? DefaultAbsoluteExpireTime ?? DefaultSlidingExpireTime;
+                foreach (var pair in redisPairs)
+                {
+                    _database.KeyExpire(pair.Key, expireTime);
```

Comment:
1. This can still leak cache keys if any call fails.
2. Actually, the existing implementation of `absoluteExpireTime` and `DefaultAbsoluteExpireTime` is incorrect, as absolute should mean `DateTime` or `DateTimeOffset`. Now both of these are effectively sliding expire time.
   - Your implementation (after adding try-catch) may be suitable to support `absoluteExpireTime` and `DefaultAbsoluteExpireTime` of the appropriate type. However, changing the parameter and property types would be a breaking change.

Let's loop the single-key `Set`.

Reply:
i don't think looping the single-key `Set` is suitable. For example, even if expire time is not give, we would still fallback to default expire time. In this case, we will always loop the single key.

Reply:
> In this case, we will always loop the single key.

What does this mean?

> Unless we value expire time consistency for each keys over the advantage of multi set key ?

Yes. From the docstrings of `IAbpCache<TKey, TValue>`'s`Set` methods:

```c#
/// Use one of the expire times at most (<paramref name="slidingExpireTime"/> or <paramref name="absoluteExpireTime"/>).
/// If none of them is specified, then
/// <see cref="ICacheOptions.DefaultAbsoluteExpireTime"/> will be used if it's not null. Othewise, <see cref="ICacheOptions.DefaultSlidingExpireTime"/>
/// will be used.
```

Reply:
> >In this case, we will always loop the single key.
>
> What does this mean?

The approach here is to loop single key `Set` conditionally right? if that is the case, to ensure that we always set expire time for each key we set, it will be something like this?
```
if any expire time given
  loop single-key "Set()"
else
  multi-key "Set()"
```

Reply:
No. Please read the quoted docstring.

Reply:
yup i understood what the quoted docstring is referring to.

so according the doc, we will always have expire time either from the parameters or from the configuration right?

in that case, we will need to change replace the implementation of using Redis String Multi Set with looping Redis String Single Key right?

Reply:
Yup, right.

Reply:
ok i will try to make the changes accordingly.

Anyway, the reason that sliding/absolute time are treated as the same expire time was due to Redis API. StringSet() only allows sliding time, if we want to support absolute time correctly, we will to use KeyExpire() API

Reply:
Yes, but the type is already incorrect in the interface and base regardless of Redis API.

Reply:
Yup, in memory cache, the workaround was to get the absolute time relative to current time.
`DateTimeOffset.Now.Add(DefaultAbsoluteExpireTime.Value));`